compute/hpc_backend.go

```diff
@@ -14,45 +15,95 @@ import (
 	"text/template"
 )
 
-// NewHPCBackend returns a HPCBackend instance.
-func NewHPCBackend(name string, submit string, conf config.Config, template string) *HPCBackend {
-	return &HPCBackend{name, submit, conf, template}
-}
-
 // HPCBackend represents an HPCBackend such as HtCondor, Slurm, Grid Engine, etc.
 type HPCBackend struct {
-	name     string
-	submit   string
-	conf     config.Config
-	template string
+	Name      string
+	SubmitCmd string
+	CancelCmd string
+	Template  string
+	Conf      config.Config
+	Event     events.Writer
+	Database  tes.ReadOnlyServer
+	// ExtractID is responsible for extracting the task id from the response
+	// returned by the SubmitCmd.
+	ExtractID func(string) string
 }
 
 // WriteEvent writes an event to the compute backend.
 // Currently, only TASK_CREATED is handled, which calls Submit.
 func (b *HPCBackend) WriteEvent(ctx context.Context, ev *events.Event) error {
 	switch ev.Type {
 	case events.Type_TASK_CREATED:
-		return b.Submit(ev.GetTask())
+		return b.Submit(ctx, ev.GetTask())
+
+	case events.Type_TASK_STATE:
+		if ev.GetState() == tes.State_CANCELED {
+			return b.Cancel(ctx, ev.Id)
+		}
 	}
 	return nil
 }
 
-// Submit submits a task via "qsub"
-func (b *HPCBackend) Submit(task *tes.Task) error {
+// Submit submits a task via "qsub", "condor_submit", "sbatch", etc.
+func (b *HPCBackend) Submit(ctx context.Context, task *tes.Task) error {
 	submitPath, err := b.setupTemplatedHPCSubmit(task)
 	if err != nil {
 		return err
 	}
 
-	cmd := exec.Command(b.submit, submitPath)
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd := exec.Command(b.SubmitCmd, submitPath)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
 	err = cmd.Run()
 	if err != nil {
+		b.Event.WriteEvent(ctx, events.NewState(task.Id, tes.SystemError))
+		b.Event.WriteEvent(
+			ctx,
+			events.NewSystemLog(
+				task.Id, 0, 0, "error",
+				"error submitting task to "+b.Name,
+				map[string]string{"error": err.Error(), "stderr": stderr.String(), "stdout": stdout.String()},
+			),
+		)
 		return err
 	}
 
-	return nil
+	backendID := b.ExtractID(stdout.String())
+
+	return b.Event.WriteEvent(
+		ctx,
+		events.NewMetadata(task.Id, 0, map[string]string{fmt.Sprintf("%s_id", b.Name): backendID}),
+	)
+}
+
+// Cancel cancels a task via "qdel", "condor_rm", "scancel", etc.
+func (b *HPCBackend) Cancel(ctx context.Context, taskID string) error {
+	var task *tes.Task
+	var err error
+
+	task, err = b.Database.GetTask(
+		ctx, &tes.GetTaskRequest{Id: taskID, View: tes.TaskView_FULL},
+	)
+	if err != nil {
+		return err
+	}
+
+	// only cancel tasks in a QUEUED state
+	state := task.GetState()
+	if state != tes.State_QUEUED {
+		return nil
+	}
+
+	backendID := getBackendTaskID(task, b.Name)
+	if backendID != "" {
+		cmd := exec.Command(b.CancelCmd, backendID)
+		return cmd.Run()
+	}
+
+	return fmt.Errorf("failed to get %s_id for task %s", b.Name, taskID)
```

Comment:
nit: the pattern I like is that exceptional conditions, such as an empty ID, are handled in the body of the if-statement, and the regular flow stays at the root level. The code here is the opposite.